akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +700,533 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+   *   been removed (canceled).
+   * @param bufferSize Total number of elements that can be buffered. If this buffer is full, the producer
+   *   is backpressured.
+   */
+  def sink[T](partitioner: (Int, T) ⇒ Int, startAfterNbrOfConsumers: Int,
+              bufferSize: Int = defaultBufferSize): Sink[T, Source[T, NotUsed]] =
+    statefulSink(() ⇒ (info, elem) ⇒ info.consumerIds(partitioner(info.size, elem)), startAfterNbrOfConsumers, bufferSize)
+
+  @DoNotInherit trait ConsumerInfo extends akka.stream.javadsl.PartitionHub.ConsumerInfo {
+
+    /**
+     * Identifiers of current consumers
+     */
+    def consumerIds: Array[Long]
+
+    /**
+     * Approximate number of buffered elements for a consumer.
+     * Larger value than other consumers could be an indication of
+     * that the consumer is slow.
+     *
+     * Note that this is a moving target since the elements are
+     * consumed concurrently.
+     */
+    def queueSize(consumerId: Long): Int
+
+    /**
+     * Number of attached consumers.
+     */
+    def size: Int
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] object Internal {
+    sealed trait ConsumerEvent
+    case object Wakeup extends ConsumerEvent
+    final case class HubCompleted(failure: Option[Throwable]) extends ConsumerEvent
+    case object Initialize extends ConsumerEvent
+
+    sealed trait HubEvent
+    case object RegistrationPending extends HubEvent
+    final case class UnRegister(id: Long) extends HubEvent
+    final case class NeedWakeup(consumer: Consumer) extends HubEvent
+    final case class Consumer(id: Long, callback: AsyncCallback[ConsumerEvent])
+    case object TryPull extends HubEvent
+
+    case object Completed
+
+    sealed trait HubState
+    final case class Open(callbackFuture: Future[AsyncCallback[HubEvent]], registrations: List[Consumer]) extends HubState
+    final case class Closed(failure: Option[Throwable]) extends HubState
+
+    private val FixedQueues = 128
+
+    // Need the queue to be pluggable to be able to use a more performant (less general)
+    // queue in Artery
+    trait PartitionQueue {
+      def init(id: Long): Unit
+      def totalSize: Int
+      def size(id: Long): Int
+      def isEmpty(id: Long): Boolean
+      def nonEmpty(id: Long): Boolean
+      def offer(id: Long, elem: Any): Unit
+      def poll(id: Long): AnyRef
+      def remove(id: Long): Unit
+    }
+
+    object ConsumerQueue {
+      val empty = ConsumerQueue(Queue.empty, 0)
+    }
+
+    final case class ConsumerQueue(queue: Queue[Any], size: Int) {
+      def enqueue(elem: Any): ConsumerQueue =
+        new ConsumerQueue(queue.enqueue(elem), size + 1)
+
+      def isEmpty: Boolean = size == 0
+
+      def head: Any = queue.head
+
+      def tail: ConsumerQueue =
+        new ConsumerQueue(queue.tail, size - 1)
+    }
+
+    class PartitionQueueImpl extends PartitionQueue {
+      private val queues1 = new AtomicReferenceArray[ConsumerQueue](FixedQueues)
```

Comment:
A design goal is to be able to use this in a allocation free mode in Artery, so that's why the queue impl is pluggable. That is most important for me.

The reason for the two implementations here is that the common case (as I see it) is to have a few (< 100) consumers over the lifetime of the hub but we must of course also support more.

Finding one single implementation would of course be better, but the difference ConcurrentHashMap  2.6 M/s vs FixedQueues 4.8 M/s (+ allocation rate consideration) might be too big difference to just pick the ConcurrentHashMap.